src/sparse_matrix.cpp

```diff
@@ -86,10 +94,23 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
 
     unsigned k = p_[i];
     unsigned row_end = p_[i + 1];
+    unsigned mid = k;
+ 
+    while (k < row_end) {
+        mid = (k + row_end)/2;
+        if (j_[mid] == j) {
+            k = mid;
+            break;
+        } else if (j_[mid] < j){
+            k = mid + 1;
+        } else {
+            row_end = mid - 1;
+        }
+    }
 
-    // TODO: Use binary search
-    while (k < row_end && j_[k] < j)
-        k++;
+    if (j_[k] < j) {
+        k = row_end;
+    }
 
     if (neq(e, zero)) {
         if (k < row_end && j_[k] == j) {
```

Comment:
Here `row_end` should be replaced with `p_[i+1]` since we changed `row_end` in the binary search